### ISSUE TITLE
Variables $(aps:slide_number)/$(aps:slides_count) are sometimes empty

### ISSUE THREAD

**Issue by Suenee** [bitfocus/companion-module-presentationtools-aps#14]
The $(aps:slide_number)/$(aps:slides_count) variables are sometimes empty even when the slide show is running.
Unfortunately, I haven't discovered the exact circumstances under which this error occurs.

OS: Win10
Companion 3.3.0 (last version)
PowerPoint 2013
APS last version

Basic stepes:
1. open APS
2. go to tab PRESENTATIONS in APS
3. choice on first position PPTX file as single presentation in APS
4. Use companion to open first presentation
5. Make check that PREVIOUS / NEXT buttons are working well
6. check the variables - sometimes are empty: $(aps:slide_number)/$(aps:slides_count)

Temporary solution:
- Restart APS and set PPTX anew
- In the middle of the presentation, the numbers disappeared again and returned after a while.

Useful information:
1. During an error, the buttons in Companion still work and APS command can be given to go to previous / next slide.
2. The "Close This" button also sometimes does not respond, although the transition between slides is still functional.

**Comment by Suenee** [comment]
I figured out why sometimes variables get emptied. This happens whenever the PowerPoint window stops being the primary active window.

Reconstruction of the sitation:
1. Open your presentation using APS (PowerPoint is primarily the active window)
2. click/switch via ALT+TAB to another application (PP will stop being the active window and APS will fail. More precisely, variables will stop working.)

Emergency solution - more options:
* Activate the APS presentation again.
* Click the mouse on the PowerPoint window
* Switch to PowerPoint with ALT+TAB
* Use the AutoHotKey script to create a hotkey that makes the window primary. Then call the defined shortcut via Vicreo. This will make the window primary again.

Still, it would be great if variables worked without these complications. That is, without the need for PowerPoint to be the primary/active window. Switching between slides works even when the window is NOT primary.

**Comment by MortenBrekkeStensland** [comment]
This is Morten from PresentationTools A/S. Thank you for your detailed report. 

As of now, PowerPoint needs to be the frontmost app for the slide-feedback to work. (The frontmost app is the one that will react to keyboard-strokes) There are a couple of reasons for this:
- APS can also control PDF and Keynote, along with PowerPoint. There could be edge-cases where more than one app is opened and displaying slides, and then our slide-feedback should apply to only the frontmost app. 
- In a lot of situations the workflow will demand PowerPoint to be the frontmost app. Then the presentation-clicker can advance the slides, and all animations will work. (some animations and videos demand that PowerPoint is in focus)
- The slide-feedback can act as a cue to confirm that PowerPoint is currently in focus. If its not displayed, you know that something is wrong.

For the future I think we can change this behaviour, at least for PC, where the slide-feedback only works for PowerPoint. But then I think we should also add a separate variable/feedback to show if PowerPoint is currently in focus. Perhaps we should add another command to force PowerPoint to be in focus too.
So, to summarise, what you discovered is in not a bug, but intended behaviour. I agree that it could be changed, but then we will need to add more features too. I can promise that we will look into this in the near future.

**Comment by Suenee** [comment]
Thanks for feedback. It would certainly be good if APS distinguished between the situation when several applications are active that APS can control at the same time (there it is necessary to clearly determine who is the main application now) and the situation when e.g. PowerPoint is alone. In case there are multiple applications running (PP, PDF, Keynote, etc.) then I agree that APS should be able to determine which one is primary. More specifically, there should be commands to specify this, otherwise confusion will result.

So far, I have bypassed it via AutoHotKey, which forces the PP to be primary whenever I move the slide.

**Comment by Suenee** [comment]
It would certainly be good if APS distinguished two cases:
1. when only PowerPoint is running (add option in settings: PP always on top)
2. there are more applications that APS controls (PowerPoint, Adobe Acrobat, Keynote...). APS can determine with a command from Companion which application is the main one on top.

I personally only use PP and have solved the problem for now by using AuthoHotKey which forces PP to be moved to the first position between the windows whenever I move the slides.

**Comment by MortenBrekkeStensland** [comment]
In Companion v.4 and above we have solved this issue for Windows: 

- New slide and build variables that only works for PowerPoint, regardless if the powerpoint presentation is in window-focus or not: 
Powerpoint_slide_number
Powerpoint_slides_count
Powerpoint_Slides_builds_count
Powerpoint_Slides_current_build

- This comes to addition to the old variables which work for both PowerPoint and PDF-presentations in Okular, but here the presentation needs to be in window-focus for these to be displayed: 
slide_number
slides_count
Slides_current_build
slides_count

**Comment by Suenee** [comment]
> * Powerpoint_slide_number
>   Powerpoint_slides_count
>   Powerpoint_Slides_builds_count
>   Powerpoint_Slides_current_build

Yes, it works better.

I found another bug with these variables. When you use:
`aps: Slide: Go to slide`

these variables does not updates. You can try:

1. Go to for example to slide number 5
2. Use `aps: Slide: Go to slide` and set the Slide Nr. 1
3. You will see, that slide changed but the variable has still number 5.

Thx

**Comment by MortenBrekkeStensland** [comment]
This issue is solved and should be closed
